aspnetcore/mvc/models/validation.md

```diff
@@ -47,7 +47,7 @@ Simply reading through the model reveals the rules about data for this app, maki
 
 * `[RegularExpression]`: Validates that the data matches the specified regular expression.
 
-* `[Required]`: Makes a property required.
+* `[Required]`: Makes a property required. Non-nullable [value types](/dotnet/csharp/language-reference/keywords/value-types) (such as `decimal`, `int`, `float`, and `DateTime`) are inherently required and don't need the `Required` attribute. If you create a [nullable type](/dotnet/csharp/programming-guide/nullable-types/) (for example `decimal?`), you can provide a functional `Required` attribute.
```

Comment:
This is very close but reads a bit like there's something wrong with `[Required]` or how MVC invokes that attribute.

Suggest:
- Mention `[Required]` can be used to change the client-side validation error message for a non-`Nullable` value type property.
- Mention `[BindRequired]` may be useful to ensure form data was complete.
  - This is an MVC attribute, not from the `System.ComponentModel.DataAnnotations` namspsace. And, just link to the correct material if it's already documented.
- Reword the last sentence to something more about expectations e.g. "The `Required` attribute works exactly as most users expect when associated with a `Nullable` value type, a `string`, or another `class` type."

----

If you need more background, here's a few server-side details:
- `[Required]` _never_ does anything server-side with non-`Nullable` value types. That's just how the attribute works.
- The MVC model binding system rejects a submission containing an empty string for a non-`Nullable` value type. But, it usually ignores completely-missing data for these properties.
  - The completely-missing case is very unusual when handling HTML form submissions: Empty data means a field was left empty while missing means the field did nothing. (Even a checkbox always submits something due to the hidden field MVC generates.)
- User's can add `[BindRequired]` to properties of non-`Nullable` value types if they want to ensure data is submitted, covering the above corner case.. (Not sure whether `[BindRequired]` is documented somewhere. It should be.)

And, a few client-side details:
- For historical reasons, `[Required]` is enforced client-side. But, the same validation is implied for all properties of non-`Nullable` value types.
- This means `[Required]` mostly allows users to change the error message shown for a client-side validation failure.

Reply:
Thanks @dougbu. It looks like this content should be in a section of the topic after all.

I'll update Tuesday afternoon.

Reply:
Up to you how much of the background needs to be documented. And, the note / section choice is up to you and @scottaddie.

Reply:
> The Required attribute works exactly as most users expect when associated with a Nullable value type, a string, or another class type.

I don't want to assume the reader's expectations or knowledge about the behavior of the attribute for a nullable type. That's why I declare that if one *creates a nullable type* that the `Required` attribute will work. However on the next commit, I intend to re-phrase that. Let's see if the next version of that sentence works.